cpp/src/arrow/compute/exec/exec_plan.h

```diff
@@ -184,6 +184,13 @@ class ARROW_EXPORT ExecNode {
   // - A method allows passing a ProductionHint asynchronously from an output node
   //   (replacing PauseProducing(), ResumeProducing(), StopProducing())
 
+  /// \brief Steps performed immediately before StartProducing is called
+  ///
+  /// This hook performs any actions in between creation of ExecPlan and the call to
+  /// StartProducing. An example could be Bloom filter pushdown. The order of ExecNodes
+  /// that executes this method is undefined, but the calls are made synchronously.
+  virtual Status PrepareToProduce() { return Status::OK(); }
```

Comment:
Can you explain why the work that goes in here can't be done as part of the node's constructor / factory function?

Reply:
When a node is constructed, the plan isn't necessarily finalized - the inputs might not be created. We need a hook in between when the plan is finalized and when we actually start running it.